### ISSUE TITLE
CSS Module Hook: Flash of Unstyled Content

### ISSUE THREAD

**Issue by seansean11** [electrode-io/electrode#190]
We're currently writing CSS for an Electrode app that takes advantage of the included postcss webpack plugins.  For example, we use quite a few [postcss-import](https://github.com/postcss/postcss-import) `@import` statements  and a lot of simple nesting, provided by [postcss-cssnext](https://github.com/MoOx/postcss-cssnext)
```css
.Page {
  &_content {
    ...
  }
}
```
I recently noticed that there is a flash of unstyled content at the beginning of every page load.  This seems to be due to the fact that css-modules-require-hook on the server doesn't know how to handle the postcss syntax with its default setup in `server->index.js`.  
```js
supports.cssModuleHook({
  generateScopedName: "[name]__[local]___[hash:base64:5]"
});
```
The HTML delivered from the server does not have any classnames for nested classes, so they are generated on the client.  For the split second before they are generated, you see a flash of unstyled content.

I tried supplying the postcss plugins directly to css-modules-require-hook, as I saw done in their documentation, and also found out it only accepts an [older version](https://github.com/rxlabs/tasty-pokedex/issues/11) of postcss-import. 
```js
supports.cssModuleHook({
  use: [
    require("postcss-import"),
    require("postcss-cssnext")
  ]
});
```
When running the app with this setup, I get no errors, but the class attributes are still not getting generated on the server-side, rather appearing after the html loads on the client.

Has anyone else dealt with this and figured out a solution?  For the time being I'm wrapping the problematic unstyled content with the [above-the-fold-only-server-render](https://github.com/electrode-io/above-the-fold-only-server-render) and not rendering it on the server altogether.

**Comment by citelao** [comment]
Is this specific to `@import`? I am having the same problem.

**Comment by seansean11** [comment]
@citelao we noticed the issue first with nested selectors not getting class names generated on the server.
```css
.Page {
  &_content {
    ...
  }
}
```
I then looked into the css-modules-require-hook and it seems as though we should be registering our postcss-cssnext plugin here, or else it won't know how to interpret the nested syntax for SSR.  After registering the postcss-cssnext plugin like in my above comment, the server actually displayed errors saying that it didn't recognize variable names that we were using (they were being imported).  Since the CSS webpack is using postcss-import I added that to my CSS module hook as well.  When I did this, the errors disappeared (and I thought everything might work), but the class names still did not generate on the server, leaving us with the same flash of unstyled content issue.

**Comment by seansean11** [comment]
@citelao Ok, with a little more tinkering I believe I found a solution:
```js
supports.cssModuleHook({
  prepend: [
    require("postcss-import"),
    require("postcss-cssnext")
  ]
});
```
This setup seemed to solve the issue for me.  The only caveat here is that you must use the 7.* version of postcss-import, as it seems all of the newer versions are async and incompatible with css-modules-require-hook.  Using `prepend` rather than `use` seemed to get this working correctly.

**Comment by mynameisrufus** [comment]
Where is supports? I have to use the following:

```javascript
const support = require("electrode-archetype-react-app/support");

support.load({
  generateScopedName: "[name]__[local]"
}).then(() => {

})
```

`generateScopedName` actually does not work so I had to monkey patch the `webpack/partials/extract.js` partial, we wanted to have the hash on the classname but they do not match from server to client.

**Comment by seansean11eis** [comment]
@mynameisrufus [supports](https://github.com/electrode-io/electrode/blob/master/packages/electrode-archetype-react-app/supports.js) looks like it points at the same file as "support".  It is most likely an artifact of a previous version.  If I remember correctly, removing the `generatedScopedName` achieved parity between the classnames on the server and client (as long as they weren't using post-css nesting or imports).  Then we needed to add those post-css plugins in order for rest of the classnames to be generated correctly.

**Comment by seansean11** [comment]
@mynameisrufus [supports](https://github.com/electrode-io/electrode/blob/master/packages/electrode-archetype-react-app/supports.js) looks like it points at the same file as "support".  It is most likely an artifact of a previous version.  If I remember correctly, removing the `generatedScopedName` achieved parity between the classnames on the server and client (as long as they weren't using post-css nesting or imports).  Then we needed to add those post-css plugins in order for rest of the classnames to be generated correctly.

**Comment by mynameisrufus** [comment]
Ok so I tracked it down to https://github.com/electrode-io/electrode/blob/master/packages/electrode-archetype-react-app/support/index.js#L16 the result of `Path.resolve(process.cwd(), "client")` is `/User/Projects/myproject/client` but our code is in `/User/Projects/myproject/src/client`.

This issue helped https://github.com/css-modules/css-modules-require-hook/issues/81

The fix for https://github.com/electrode-io/electrode/blob/master/packages/electrode-archetype-react-app-dev/config/webpack/partial/extract-style.js was:

```js
return mergeWebpackConfig(config, {
   plugins: [
     new webpack.LoaderOptionsPlugin({
       options: {
         context: Path.resolve(process.cwd(), "client")
       }
    }
  ]
}
```

The reason the code is in `src` was because of following the `yo` generated example.

**Comment by karlhorky** [comment]
@mynameisrufus I've implemented this fix in a new pull request here: #258.
